### PR TITLE
OLMIS-6614: Added serving of facilityId and programId params and reduced number of queries to DB made by `GET /api/orderableFulfills` endpoint

### PR DESCRIPTION
**Pull request by dserkowski**

### CODE REVIEW

File: src/main/java/org/openlmis/referencedata/service/OrderableService.java [hunk 1]
```diff
@@ -50,10 +48,6 @@
       Pageable pageable) {
     LOGGER.info("search orderable query params: {}", queryMap);
 
-    if (!queryMap.isValid()) {
```
Comment (lukaslew) [inline src/main/java/org/openlmis/referencedata/service/OrderableService.java]: why this has been removed?

File: src/main/java/org/openlmis/referencedata/service/OrderableService.java [hunk 2]
```diff
@@ -87,11 +81,6 @@
   public List<Orderable> getLatestLastUpdatedDate(@NotNull QueryOrderableSearchParams queryMap,
       Pageable pageable) {
 
-    if (!queryMap.isValid()) {
```
Comment (lukaslew) [inline src/main/java/org/openlmis/referencedata/service/OrderableService.java]: why we remove this check?

File: src/main/java/org/openlmis/referencedata/util/EntityCollection.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+package org.openlmis.referencedata.util;
+
+import com.google.common.collect.Maps;
+import java.util.Collection;
+import java.util.Map;
+import java.util.UUID;
+import org.openlmis.referencedata.domain.BaseEntity;
+
+public class EntityCollection<T extends BaseEntity> {
```
Comment (lukaslew) [inline src/main/java/org/openlmis/referencedata/util/EntityCollection.java]: can we explain why this class should be used and when?

Comment (lukaslew) [reply src/main/java/org/openlmis/referencedata/util/EntityCollection.java]: do we really need this class?

Comment (dserkowski) [reply src/main/java/org/openlmis/referencedata/util/EntityCollection.java]: Thanks this we can reduce number of calls to DB and (thanks for applying a hashMap) we can fetch entity in O(1) time. 
This is used in places when we need to findAll entities and at the same time find an entry (of already fetched entities) by id.

File: src/main/java/org/openlmis/referencedata/service/OrderableService.java (hunk 2 above)
Comment (dserkowski) [reply src/main/java/org/openlmis/referencedata/service/OrderableService.java]: I have moved it constructor of QueryOrderableSearchParams.
https://github.com/OpenLMIS/openlmis-referencedata/blob/a66450c22c27cb1dec14399d5cb2c8a5c058089c/src/main/java/org/openlmis/referencedata/web/QueryOrderableSearchParams.java#L55

File: src/main/java/org/openlmis/referencedata/service/OrderableService.java (hunk 1 above)
Comment (dserkowski) [reply src/main/java/org/openlmis/referencedata/service/OrderableService.java]: I have moved it to the constructor of QueryOrderableSearchParams.
https://github.com/OpenLMIS/openlmis-referencedata/blob/a66450c22c27cb1dec14399d5cb2c8a5c058089c/src/main/java/org/openlmis/referencedata/web/QueryOrderableSearchParams.java#L55

File: src/main/resources/api-definition.yaml [hunk 4]
```diff
@@ -3321,10 +3321,25 @@ resourceTypes:
...
-                  description: Orderable id
+                  description: >
+                      Orderable id. Cannot be provided together with facilityId and programId.
                   type: string
                   required: false
                   repeat: true
+              facilityId:
+                  description: >
+                      Facility Id. This param allows filtering the result. If provided, the
+                      programId param also must be provided.
```
Comment (sebbrudzinski) [inline src/main/resources/api-definition.yaml]: Worth explaining how exactly this filtering works. OrderableFulfills resource doesn't have a facility or a program - how is it filtered then? Explain that this employs FTAPs, that the FTAP must be active, etc.

File: src/main/java/org/openlmis/referencedata/util/messagekeys/OrderableFulFillMessageKeys.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details. You should have received a copy of
+ * the GNU Affero General Public License along with this program. If not, see
+ * http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org.
+ */
+
+package org.openlmis.referencedata.util.messagekeys;
+
+public abstract class OrderableFulFillMessageKeys extends MessageKeys {
```
Comment (mkwiatkowskisoldevelo) [inline src/main/java/org/openlmis/referencedata/util/messagekeys/OrderableFulFillMessageKeys.java]: Please add messages associated with those keys.

File: src/main/resources/api-definition.yaml (hunk 4 above)
Comment (dserkowski) [reply src/main/resources/api-definition.yaml]: Fixed

**Pull request merged**